### ISSUE TITLE
Removing TapGestureRecognizer and then re-add it inside its command throws an Exception

### ISSUE THREAD

**Issue by codercampos** [xamarin/Xamarin.Forms#2021]
### Description
We have found a particular issue when you try to remove a TapGestureRecognizer from a element inside its command and then re-add it to that element. When tries to add it again, throws an InvalidOperationException. This only happens on Android, on iOS seems to work without problem.

### Steps to Reproduce

1. Set a new TapGestureRecognizer
2. Set the a new Command
3. In the Command, remove the TapGestureRecognizer from the collection you will add it
4. Execute any code you want (maybe a new Page call)
5. Re-Add the TapGestureRecognizer to the control you previously added

Example code:
```
var tabGesture = new TapGestureRecognizer();
var animateCommand = new Command(async () =>
{
      stack.GestureRecognizers.Remove(tabGesture);
      await stack.ScaleTo(.96, 100);
      await stack.ScaleTo(1, 100);
      DataContext.SearchPopupCommand.Execute(key);
      stack.GestureRecognizers.Add(tabGesture);
});
tabGesture.Command = animateCommand;
stack.GestureRecognizers.Add(tabGesture);
```

### Expected Behavior
Command should be executed without problem.

### Actual Behavior
On iOS: Works fine. Command is executed.
On Android: System.InvalidOperationException is thrown: Collection was modified; enumeration operation may not execute.

### Basic Information

- Version with issue: Version 2.5.0.122203
- Last known good version: Unknown 
- IDE: Visual Studio for Mac 7.3.3
- Platform Target Frameworks: 
  - iOS:  11.6.1.4 (with Xcode 9.2)
  - Android: 7.1
- Android Support Library Version: 25.4.0.2 (Xamarin.Android.Support.v7.AppCompat)
- Affected Devices: Nexus 5 Android Emulator with Google Services (Android 5.0), Samsung Core 2 (SM-G355M and Android 4.4.2)

**Comment by hartez** [comment]
This also crashes on iOS and UWP.

**Comment by hartez** [comment]
[Repro](https://github.com/xamarin/Xamarin.Forms/files/1786626/GitHub2021.Repro.zip)

**Comment by christoroth** [comment]
I have this issue.  It can be simplified down to the remove causing the issue.  It occurs on UWP and Android but not iOS with Xamarin Forms 3.  Removing the gesture recognizer during the handler for the gesture causes the "collection modified" error, I presume because the collection of gestures is being iterated to call the handlers.  

In my case I was doing a one shot tap to allow the user to tap outside of a popup element to trigger its removal (press button to add 'blanket' and popup, adds tap gesture to blanket, tap on blanket removes gesture, then removes popup and popup).

It's a bit messy but I have worked around it temporarily by allowing my handler to complete and run my logic in a separate thread by wrapping it within my handler: 

var t = Task.Run(delegate
            {
                Device.BeginInvokeOnMainThread(() =>
                {
                          ... do my ui changes - remove gesture recogniser, blanket, popup etc
                });
             });

**Comment by Maxima078** [comment]
I had a similar issue removing a PanGesture from GestureRecognizers list **within** the PanUpdated handler using Xamarin.Forms 3.1.0.550168-pre3. Updating to 3.1.0.583944 fixed my issue but I couldn't find the commit that fixed it, can someone point me to the right direction?

@christoroth this is not a thread problem, your workaround fix it because the main thread will execute code after you have jumped out from your handler. You could just fire and forget a Task.Run with a short delay before removing your recogniser. Furthermore in your workaround you do not need the Task.run.

**Comment by codercampos** [comment]
Maybe add the Handler collection besides the main collection could fix this. I don't if this is a good practice (I think it isn't ) but maybe this enlights someone else

**Comment by AmrAlSayed0** [comment]
If you want to make sure that the command isn't fired while the element is being animated (or whatever it is doing) then you can use then you can provide a function for the "CanExcute" and call Command.ChangeCanExcute() when you want the command to recheck its status.

```csharp
var tabGesture = new TapGestureRecognizer();
var isAnimating;
var animateCommand = new Command(async () =>
{
      isAnimating = true;
      tabGesture.Command?.ChangeCanExcute();
      await stack.ScaleTo(.96, 100);
      await stack.ScaleTo(1, 100);
      DataContext.SearchPopupCommand.Execute(key);
      isAnimating = false;
      tabGesture.Command?.ChangeCanExcute();
},
() => !isAnimating );

tabGesture.Command = animateCommand;
stack.GestureRecognizers.Add(tabGesture);
```

**Comment by PureWeen** [comment]
This looks to be fixed in 3.1

If that seems incorrect we can reopen
